vms/platformvm/state/state.go

```diff
@@ -1994,3 +2054,105 @@ func (s *state) writeMetadata() error {
 	}
 	return nil
 }
+
+// Returns the block, status of the block, and whether it is a [stateBlk].
+// Invariant: blkBytes is safe to parse with blocks.GenesisCodec
+//
+// TODO: Cleanup after v1.11.x is activated
+func parseStoredBlock(blkBytes []byte) (blocks.Block, choices.Status, bool, error) {
+	blk, err := blocks.Parse(blocks.GenesisCodec, blkBytes)
+	if err == nil {
+		return blk, choices.Accepted, false, nil
+	}
+
+	// Fallback to [stateBlk], we have not finished pruning.
+	blkState := stateBlk{}
+	if _, err := blocks.GenesisCodec.Unmarshal(blkBytes, &blkState); err != nil {
+		return nil, choices.Processing, false, err
+	}
+
+	blkState.Blk, err = blocks.Parse(blocks.GenesisCodec, blkState.Bytes)
+	if err != nil {
+		return nil, choices.Processing, false, err
+	}
+
+	return blkState.Blk, blkState.Status, true, nil
+}
+
+func (s *state) PruneAndIndex() error {
+	lastAcceptedBlk, err := s.GetStatelessBlock(s.lastAccepted)
+	if err != nil {
+		return fmt.Errorf("failed to get last accepted block from local disk: %w", err)
+	}
+
+	lastAcceptedHeight := lastAcceptedBlk.Height()
+	lastAcceptedHeightKey := database.PackUInt64(lastAcceptedHeight)
+	if _, err := database.GetID(s.blockIDDB, lastAcceptedHeightKey); err == nil {
+		// If the last accepted block is in [s.blockIDDB], we have previously
+		// run this function.
+		return nil
+	}
+
+	// It is possible that new blocks are added after grabbing this iterator. New
+	// blocks are guaranteed to be accepted and height-indexed, so we don't need to
+	// check them.
+	blockIterator := s.blockDB.NewIterator()
+	defer blockIterator.Release()
+
+	s.ctx.Log.Info("populating platformvm block height index")
+	var (
+		startTime  = time.Now()
+		lastUpdate = startTime
+	)
+
+	for indexed := uint64(0); blockIterator.Next(); {
+		blkBytes := blockIterator.Value()
+
+		blk, status, isStateBlk, err := parseStoredBlock(blkBytes)
+		if err != nil {
+			return err
+		}
+
+		if status != choices.Accepted {
+			// We no longer persist any non-accepted blocks to disk.
+			if err := s.blockDB.Delete(blockIterator.Key()); err != nil {
+				return fmt.Errorf("failed to delete block: %w", err)
+			}
+		}
```

Comment:
```suggestion
			}
			continue
		}
```
I don't think we want to index the height of non-accepted blocks